tests/metrics/test_computed_metrics.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+from typing import Tuple
+from unittest.mock import MagicMock
+
+import pytest
+
+from composer.core import Callback, State
+from composer.core.types import DataLoader
+from composer.loggers import Logger
+from composer.trainer import Trainer
+from tests.fixtures.models import SimpleBatchPairModel
+
+
+class TestMetricsCallback(Callback):
+
+    def __init__(self, compute_training_metrics: bool, compute_val_metrics: bool) -> None:
+        self.compute_training_metrics = compute_training_metrics
+        self.compute_val_metrics = compute_val_metrics
+        self._train_batch_end_train_accuracy = None
+        self._eval_batch_end_accuracy = None
+
+    def init(self, state: State, logger: Logger) -> None:
+        # on init, the `computed_metrics` should be empty
+        del logger  # unused
+        assert state.computed_metrics == {}, "no metrics should be defined on init()"
+
+    def batch_end(self, state: State, logger: Logger) -> None:
+        # The metric should be computed and updated on state every batch.
+        del logger  # unused
+        if self.compute_training_metrics:
+            # assuming that at least one sample was correctly classified
+            assert state.computed_metrics["train"]["Accuracy"] != 0.0
+            self._train_batch_end_train_accuracy = state.computed_metrics["train"]["Accuracy"]
+
+    def epoch_end(self, state: State, logger: Logger) -> None:
+        # The metric at epoch end should be the same as on batch end.
+        del logger  # unused
+        if self.compute_training_metrics:
+            assert state.computed_metrics["train"]["Accuracy"] == self._train_batch_end_train_accuracy
+
+    def eval_batch_end(self, state: State, logger: Logger) -> None:
+        # The validation accuracy should be defined after each eval batch
+        if self.compute_val_metrics:
+            # assuming that at least one sample was correctly classified
+            assert state.computed_metrics["eval"]["Accuracy"] != 0.0
+            self._eval_batch_end_accuracy = state.computed_metrics["eval"]["Accuracy"]
+
+    def eval_end(self, state: State, logger: Logger) -> None:
+        if self.compute_val_metrics:
+            assert state.computed_metrics["eval"]["Accuracy"] == self._eval_batch_end_accuracy
+
+
+@pytest.mark.parametrize('compute_training_metrics', [True, False])
+@pytest.mark.parametrize('validate_every_n_batches', [-1, 1])
+@pytest.mark.parametrize('validate_every_n_epochs', [-1, 1])
+def test_computed_metrics(
+    dummy_train_dataloader: DataLoader,
+    dummy_val_dataloader: DataLoader,
+    dummy_num_classes: int,
+    dummy_in_shape: Tuple[int, ...],
+    compute_training_metrics: bool,
+    validate_every_n_batches: int,
+    validate_every_n_epochs: int,
+):
+    num_channels = dummy_in_shape[0]
```

Comment:
this test could benefit from line breaks and organization to break up into paragraphs, e.g. before `Trainer(`, before `trainer.fit(`.

The code around creaing the metrics callback should also be moved above the `trainer(`, since the model and mock loggers are also created.